### PR TITLE
service privileges: API docs & completion

### PR DESCRIPTION
**Pull request by aluzzardi**
Follow up for #32339

Added Swagger Docs, updated the API changelog and shell completion.

/cc @thaJeztah @friism @mstanleyjones

### CODE REVIEW

**Comment by mstanleyjones** [comment]
I think you also need to update the command-line reference files, right?

**Comment by thaJeztah** [comment]
Yes, this needs an update to the CLI docs as well; although plans are to remove the `USAGE` output there, we probably could use a simple example of it's usage

**Comment by thaJeztah** [comment]
Added a commit for the CLI reference

/cc @friism @johnstep if the description is correct

File: docs/reference/commandline/service_create.md [hunk 1]
```diff
@@ -761,6 +762,16 @@ $ docker service create --name dns-cache -p 53:53/tcp -p 53:53/udp dns-cache
 $ docker service create --name dns-cache -p 53:53/udp dns-cache
 ```
 
+### Provide credential specs for managed service accounts (Windows only)
+
+This option is only supported on a daemon running on Windows. The `--credential-spec` must be in the format
```
Comment (mstanleyjones) [inline docs/reference/commandline/service_create.md]: Does this mean running Windows containers, or running Linux containers on a Windows host machine?

**Review by mdlinville — COMMENTED** [review]
Just one question...

File: docs/reference/commandline/service_create.md (hunk 1 above)
Comment (thaJeztah) [inline docs/reference/commandline/service_create.md]: Hm, yes, it's Windows containers; should I change to "only supported for Native Windows containers"?

Comment (friism) [inline docs/reference/commandline/service_create.md]: Yeah that, or `supported on a daemon running natively on Windows`

**Comment by aaronlehmann** [comment]
@aluzzardi: Need a rebase

**Comment by aluzzardi** [comment]
Rebased

File: docs/reference/commandline/service_create.md (hunk 1 above)
Comment (aaronlehmann) [inline docs/reference/commandline/service_create.md]: Let's remove the reference to the daemon. As I understand it, this applies to the containers being run, not the daemon that the client is communicating with to set up the service.

File: docs/reference/commandline/service_create.md [hunk 2]
```diff
@@ -761,6 +762,18 @@ $ docker service create --name dns-cache -p 53:53/tcp -p 53:53/udp dns-cache
 $ docker service create --name dns-cache -p 53:53/udp dns-cache
 ```
 
+### Provide credential specs for managed service accounts (Windows only)
+
+This option is only used for services using Windows containers. The
```
Comment (thaJeztah) [inline docs/reference/commandline/service_create.md]: @aaronlehmann @mstanleyjones does this work?

Comment (aaronlehmann) [inline docs/reference/commandline/service_create.md]: SGTM

File: docs/reference/commandline/service_create.md [hunk 3]
```diff
@@ -761,6 +762,18 @@ $ docker service create --name dns-cache -p 53:53/tcp -p 53:53/udp dns-cache
...
+`--credential-spec` must be in the format `file://spec.txt` or
+`registry://keyname`.
+
+When using the `file://` format, the referenced file must be present on the
+daemon host in the `C:\programfiles\docker\credentialspecs\` directory. When
```
Comment (aaronlehmann) [inline docs/reference/commandline/service_create.md]: Let's make clear whether the path after `file://` should be a full path or a relative path within `C:\programfiles\docker\credentialspecs\`.

As a side note, it's potentially confusing if `file://` is meant to be used with a relative path here.

Comment (thaJeztah) [inline docs/reference/commandline/service_create.md]: Yes, I was a bit confused by this as well https://github.com/docker/docker/pull/32339#discussion_r110453491

@friism could you help a bit here?

Comment (thaJeztah) [inline docs/reference/commandline/service_create.md]: also; does it accept a filename only or also a sub directory?

Comment (friism) [inline docs/reference/commandline/service_create.md]: It's a relative path from the `C:\programfiles\docker\credentialspecs` folder. Using an absolute path errors correctly:

    time="2017-04-17T08:48:45.126507000-07:00" level=error msg="fatal task error" error="starting container failed: invalid credential spec - file:// path cannot be absolute" module="node/agent/taskmanager" node.id=y6t3hh97zaqh3wkxnrnkg6plw service.id=l5bz9jvh9qxccxdladbg7uec2 task.id=0s3vbr0luxotquovs2ul3fosr

Subdirs can be used, eg. I just tested `C:\ProgramData\docker\credentialspecs\subdir` and it works great.

@aluzzardi let me know if you want to work on the wording or whether you want to do it.

**Comment by thaJeztah** [comment]
ping @aluzzardi can you address the review comments?

**Comment by LK4D4** [comment]
@thaJeztah maybe you could carry it?

**Comment by LK4D4** [comment]
ping @thaJeztah

**Comment by thaJeztah** [comment]
Carrying API side in https://github.com/moby/moby/issues/34002 - will close this one when I opened a PR for the CLI docs

**Comment by LK4D4** [comment]
@thaJeztah assigned to you

File: docs/reference/commandline/service_create.md (hunk 3 above)
Comment (johnstep) [inline docs/reference/commandline/service_create.md]: Actually it is `$env:ProgramData`, not `$env:ProgramFiles` (PowerShell environment variable syntax), so the default is `$env:ProgramData\Docker\CredentialSpecs` which is usually `C:\ProgramData\Docker\CredentialSpecs`.

Is there a better way to phrase this to say it's `CredentialSpecs` under the Docker data root directory, which defaults to `$env:ProgramData\Docker` where `$env:ProgramData` is usually `C:\ProgramData`?

File: api/swagger.yaml [hunk 4]
```diff
@@ -1939,6 +1939,39 @@ definitions:
...
+              CredentialSpec:
+                type: "object"
+                description: "CredentialSpec for managed service account (Windows only)"
+                properties:
+                  File:
+                    type: "string"
+                    description: "Load credential spec from this file"
+                  Registry:
+                    type: "string"
+                    description: "Load credential spec from this registry key"
```
Comment (johnstep) [inline api/swagger.yaml]: Technically it is a registry _value_ within the CredentialSpecs registry key, so I would change this to 'registry value'. Comparing to a file system, a key is roughly equivalent to a directory, and a value, a file.

File: docs/reference/commandline/service_create.md [hunk 5]
```diff
@@ -761,6 +762,18 @@ $ docker service create --name dns-cache -p 53:53/tcp -p 53:53/udp dns-cache
 $ docker service create --name dns-cache -p 53:53/udp dns-cache
 ```
 
+### Provide credential specs for managed service accounts (Windows only)
+
+This option is only used for services using Windows containers. The
+`--credential-spec` must be in the format `file://spec.txt` or
```
Comment (johnstep) [inline docs/reference/commandline/service_create.md]: To be clear on the expected file format, maybe change this to `file://spec.json`.

File: docs/reference/commandline/service_create.md [hunk 6]
```diff
@@ -761,6 +762,18 @@ $ docker service create --name dns-cache -p 53:53/tcp -p 53:53/udp dns-cache
...
+`registry://keyname`.
+
+When using the `file://` format, the referenced file must be present on the
+daemon host in the `C:\programfiles\docker\credentialspecs\` directory. When
+using the `registry://keyname` format, the credential spec is read from
```
Comment (johnstep) [inline docs/reference/commandline/service_create.md]: `registry://valuename`

File: docs/reference/commandline/service_create.md [hunk 7]
```diff
@@ -761,6 +762,18 @@ $ docker service create --name dns-cache -p 53:53/tcp -p 53:53/udp dns-cache
 $ docker service create --name dns-cache -p 53:53/udp dns-cache
 ```
 
+### Provide credential specs for managed service accounts (Windows only)
+
+This option is only used for services using Windows containers. The
+`--credential-spec` must be in the format `file://spec.txt` or
+`registry://keyname`.
```
Comment (johnstep) [inline docs/reference/commandline/service_create.md]: `registry://valuename`

**Review by johnstep — CHANGES_REQUESTED** [review]

File: docs/reference/commandline/service_create.md (hunk 5 above)
Comment (thaJeztah) [inline docs/reference/commandline/service_create.md]: @johnstep I was already working on the API changes in https://github.com/moby/moby/pull/34002; PTAL there

**Comment by thaJeztah** [comment]
https://github.com/docker/cli/pull/338 and https://github.com/moby/moby/pull/34002 were merged, so closing this one

**Pull request closed without merging**